### PR TITLE
Ten times faster than before in GPU get_best_span of bidaf

### PR DESCRIPTION
**Pull request by Meelfy**
old version
- CPU 34.3 ms ± 973 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
- GPU 60.2 ms ± 1.33 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

new version
- CPU 47.4 ms ± 1.45 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
- GPU 5.04 ms ± 12.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

Ten times faster than before in GPU

### CODE REVIEW

**Comment by Meelfy** [comment]
Script for testing speed：

```python
import torch
def get_best_span(span_start_logits: torch.Tensor, span_end_logits: torch.Tensor) -> torch.Tensor:
    if span_start_logits.dim() != 2 or span_end_logits.dim() != 2:
        raise ValueError("Input shapes must be (batch_size, passage_length)")
    batch_size, passage_length = span_start_logits.size()
    max_span_log_prob = [-1e20] * batch_size
    span_start_argmax = [0] * batch_size
    best_word_span = span_start_logits.new_zeros((batch_size, 2), dtype=torch.long)

    span_start_logits = span_start_logits.detach().cpu().numpy()
    span_end_logits = span_end_logits.detach().cpu().numpy()

    for b in range(batch_size):  # pylint: disable=invalid-name
        for j in range(passage_length):
            val1 = span_start_logits[b, span_start_argmax[b]]
            if val1 < span_start_logits[b, j]:
                span_start_argmax[b] = j
                val1 = span_start_logits[b, j]

            val2 = span_end_logits[b, j]

            if val1 + val2 > max_span_log_prob[b]:
                best_word_span[b, 0] = span_start_argmax[b]
                best_word_span[b, 1] = j
                max_span_log_prob[b] = val1 + val2
    return best_word_span

def fast_get_best_span(span_start_logits: torch.Tensor,
                       span_end_logits: torch.Tensor) -> torch.Tensor:
    if span_start_logits.dim() != 2 or span_end_logits.dim() != 2:
        raise ValueError("Input shapes must be (batch_size, passage_length)")
    batch_size, passage_length = span_start_logits.size()
    device = span_start_logits.device
    logits_add = span_start_logits.unsqueeze(2) + span_end_logits.unsqueeze(1)
    logits_add_triu = logits_add * torch.triu(torch.ones((passage_length, passage_length),
                                                         device=device)).unsqueeze(0)
    res = logits_add_triu.view(batch_size, -1).argmax(-1)
    span_start_idx = res // passage_length
    span_end_idx = res % passage_length
    return torch.stack([span_start_idx, span_end_idx], dim=-1)

batch_size, passage_length = 32, 500
span_start_logits = torch.rand((batch_size, passage_length))
span_end_logits = torch.rand((batch_size, passage_length))
%timeit get_best_span(span_start_logits, span_end_logits)
%timeit fast_get_best_span(span_start_logits, span_end_logits)
span_start_logits = span_start_logits.cuda(1)
span_end_logits = span_start_logits.cuda(1)
%timeit get_best_span(span_start_logits, span_end_logits)
%timeit fast_get_best_span(span_start_logits, span_end_logits)
```

**Comment by matt-gardner** [comment]
Thanks @Meelfy, I can see how this would be a lot faster.  It's failing the test `get_best_span` test, though, so it's not actually equivalent logic.  It looks like it's returning invalid spans - this logic gives `(4, 3)` instead of `(0, 0)` on the first test case, where `(4, 3)` is not a valid span.  Maybe you need `torch.tril` instead of `torch.triu`?

**Comment by matt-gardner** [comment]
Oh, no, it's probably not `torch.tril`, it's that you're multiplying by 0 trying to ignore logits.  I was wondering why this worked when I first read the logic, but assumed that it was ok, until I saw the failing test.  The logits will be negative, so multiplying by 0 / 1 won't work.  You need to take the log of the `torch.triu()` call, and add instead of multiply.

**Comment by Meelfy** [comment]
Thanks for your quick reply! I missed the situation that logits will be negative before. Because I use probs to calculate span myself.  Can you tell me how to run the test case? I am not very familiar with this. I have modified the code according to your suggestion. Thx!

**Comment by HarshTrivedi** [comment]
If you want to test everything (but it will take a lot of time): 
```
python scripts/verify.py --checks pytest
```

If you need to test only one file: 
```
pytest -v --color=yes allennlp/test/models/reading_comprehensions/bidaf_test
```

**Comment by Meelfy** [comment]
```python
def get_best_span(span_start_logits: torch.Tensor,
                       span_end_logits: torch.Tensor) -> torch.Tensor:
    if span_start_logits.dim() != 2 or span_end_logits.dim() != 2:
        raise ValueError("Input shapes must be (batch_size, passage_length)")
    batch_size, passage_length = span_start_logits.size()
    device = span_start_logits.device
    # This operation is very much like outer product.
    # So the last two dimensions of logits_add
    # logits_add[:, i, j] == start_logits[:, i] + start_logits[:, j]
    logits_add = span_start_logits.unsqueeze(2) + span_end_logits.unsqueeze(1)
    # This is a mask operation. Ignore the value in logits_add when (i < j)
    logits_add_triu = logits_add + torch.triu(torch.ones((passage_length, passage_length),
                                                         device=device)).log().unsqueeze(0)
    # Just get the max value's idx of logits_add_triu[b, :, :] for batch b 
    res = logits_add_triu.view(batch_size, -1).argmax(-1)
    span_start_idx = res // passage_length
    span_end_idx = res % passage_length
    return torch.stack([span_start_idx, span_end_idx], dim=-1
```

**Comment by Meelfy** [comment]
> If you want to test everything (but it will take a lot of time):
> 
> ```
> python scripts/verify.py --checks pytest
> ```
> If you need to test only one file:
> 
> ```
> pytest -v --color=yes allennlp/test/models/reading_comprehensions/bidaf_test
> ```



> If you want to test everything (but it will take a lot of time):
> 
> ```
> python scripts/verify.py --checks pytest
> ```
> If you need to test only one file:
> 
> ```
> pytest -v --color=yes allennlp/test/models/reading_comprehensions/bidaf_test
> ```

I am sorry to ask simple questions casually. Thank you very much for your reply.

**Comment by matt-gardner** [comment]
I normally just use the `-k` flag for pytest - a lot easier than specifying a particular file.

File: allennlp/models/reading_comprehension/bidaf.py [hunk 1]
```diff
@@ -299,27 +299,26 @@ def get_metrics(self, reset: bool = False) -> Dict[str, float]:
...
-                    best_word_span[b, 1] = j
-                    max_span_log_prob[b] = val1 + val2
-        return best_word_span
+        device = span_start_logits.device
+        # (batch_size, passage_length, passage_length)
+        span_log_probs = span_start_logits.unsqueeze(2) + span_end_logits.unsqueeze(1)
+        # Only the upper triangle of the span matrix is valid; the lower triangle has entries where
+        # the span ends before it starts.
+        span_log_mask = torch.triu(torch.ones((passage_length, passage_length),
+                                              device=device)).log().unsqueeze(0)
```
Comment (Meelfy) [inline allennlp/models/reading_comprehension/bidaf.py]: It seems like unqueeze is useless here. But it doesn't matter.
